pkg/cmd/pulumi/replay_events.go

```diff
@@ -0,0 +1,187 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	"github.com/pulumi/pulumi/pkg/v3/backend/display"
+	"github.com/pulumi/pulumi/pkg/v3/engine"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+func newReplayEventsCmd() *cobra.Command {
+	var preview bool
+
+	var jsonDisplay bool
+	var diffDisplay bool
+	var showConfig bool
+	var showReplacementSteps bool
+	var showSames bool
+	var showReads bool
+	var suppressOutputs bool
+	var debug bool
+
+	var delay time.Duration
+
+	var cmd = &cobra.Command{
+		Use:   "replay-events [kind] [events-file]",
+		Short: "Replay events from a prior update, refresh, or destroy",
+		Long: "Replay events from a prior update, refresh, or destroy.\n" +
+			"\n" +
+			"This command is used to replay events emitted by a prior\n" +
+			"invocation of the Pulumi CLI (e.g. `pulumi up --event-log [file]`).\n" +
+			"\n" +
+			"This command loads events from the indicated file and renders them\n" +
+			"using either the progress view or the diff view.\n",
+		Args:   cmdutil.ExactArgs(2),
+		Hidden: !hasDebugCommands(),
+		Run: cmdutil.RunFunc(func(cmd *cobra.Command, args []string) error {
+			var action apitype.UpdateKind
+			switch args[0] {
+			case "update":
+				if preview {
+					action = apitype.PreviewUpdate
+				} else {
+					action = apitype.UpdateUpdate
+				}
+			case "refresh":
+				action = apitype.RefreshUpdate
+			case "destroy":
+				action = apitype.DestroyUpdate
+			case "import":
+				action = apitype.ResourceImportUpdate
+			default:
+				return fmt.Errorf("unrecognized update kind '%v'", args[0])
+			}
+
+			var displayType = display.DisplayProgress
+			if diffDisplay {
+				displayType = display.DisplayDiff
+			}
+
+			displayOpts := display.Options{
+				Color:                cmdutil.GetGlobalColorization(),
+				ShowConfig:           showConfig,
+				ShowReplacementSteps: showReplacementSteps,
+				ShowSameResources:    showSames,
+				ShowReads:            showReads,
+				SuppressOutputs:      suppressOutputs,
+				IsInteractive:        cmdutil.Interactive(),
+				Type:                 displayType,
+				JSONDisplay:          jsonDisplay,
+				Debug:                debug,
+			}
+
+			events, err := loadEvents(args[1])
+			if err != nil {
+				return fmt.Errorf("error reading events: %w", err)
+			}
+
+			eventChannel, doneChannel := make(chan engine.Event), make(chan bool)
+
+			if delay != 0 {
+				time.Sleep(delay)
+			}
+
+			go display.ShowEvents(
+				"replay", action, "replay", "replay",
+				eventChannel, doneChannel, displayOpts, preview)
+
+			for _, e := range events {
+				eventChannel <- e
+			}
+			<-doneChannel
+
+			return nil
+		}),
+	}
+
+	cmd.PersistentFlags().BoolVarP(
```

Comment:
Its a bit sad this is needed but looks to be a gnarly untangling to get rid of it.

Reply:
Yeah. We might be able to scrape this out of the events if we have enough of them--I think that the resource events include a bit that indicates whether or not they're part of a preview. Feels like a better solution would be to issue a "start" event that contains this info, though.